tfjs-backend-webgpu/src/shader_preprocessor_wgsl.ts

```diff
@@ -0,0 +1,604 @@
+          return ${texName}.numbers[${
+          outRank > 1 ? 'getOutputFlatIndex(coords)' : 'coords'} / 4u];
+        }
+        `;
+    } else {
+      return `
+      fn ${funcName}ByGlobalId(globalId : vec3<u32>) -> f32 {
+        return f32(${texName}.numbers[globalId.x]);
+      }
+
+      fn ${funcName}ByCoords(coords : ${type}) -> f32 {
+        return f32(${texName}.numbers[${
+          outRank > 1 ? 'getOutputFlatIndex(coords)' : 'coords'}]);
+      }
+      `;
+    }
+  }
+
+  const broadcastDims = backend_util.getBroadcastDims(inInfo.shape, outShape);
+  const rankDiff = outRank - inRank;
+
+  let coordsSnippet = '';
+
+  if (inRank === 0) {
+    if (isVec4) {
+      return `
+      fn ${funcName}ByGlobalId(globalId : vec3<u32>) -> vec4<f32> {
+        return get${texFuncSnippet}();
+      }
+
+      fn ${funcName}ByCoords(coords : ${type}) -> vec4<f32> {
+        return get${texFuncSnippet}();
+      }
+    `;
+    }
+    return `
+      fn ${funcName}ByGlobalId(globalId : vec3<u32>) -> f32{
+        return get${texFuncSnippet}();
+      }
+
+      fn ${funcName}ByCoords(coords : ${type}) -> f32{
+        return get${texFuncSnippet}();
+      }
+    `;
+  } else {
+    if (outRank < 2 && broadcastDims.length >= 1) {
+      coordsSnippet = 'coords = 0u;';
+    } else {
+      coordsSnippet =
+          broadcastDims.map(d => `coords[${d + rankDiff}u] = 0u;`).join('\n');
+    }
+  }
+
+  let unpackedCoordsSnippet = '';
+  if (outRank < 2 && inRank > 0) {
+    unpackedCoordsSnippet = 'coords';
+  } else {
+    if (outRank > 1) {
+      const coordsType = getCoordsDataTypeWgsl(inRank);
+      const coordsValues =
+          inInfo.shape.map((s, i) => `coords[${i + rankDiff}u]`).join(', ');
+      unpackedCoordsSnippet = `${coordsType}(${coordsValues})`;
+    } else {
+      unpackedCoordsSnippet = 'coords';
+    }
+  }
+
+  const shapeStr =
+      `uniforms.${texName.charAt(0).toLowerCase() + texName.slice(1)}Shape`;
+  const rankStr = `${inRank}D`;
+  if (isVec4) {
+    return `
+      fn ${funcName}ByGlobalId(globalId : vec3<u32>) -> vec4<f32> {
+        var coords : ${type} = getOutputCoords(globalId);
+        ${coordsSnippet}
+        return ${texName}.numbers[getFlatIndex${rankStr}(${
+        unpackedCoordsSnippet}, ${shapeStr}) / 4u];
+      }
+
+      fn ${funcName}ByCoords(coordsIn : ${type}) -> vec4<f32> {
+        var coords : ${type} = coordsIn;
+        ${coordsSnippet}
+        return ${texName}.numbers[getFlatIndex${rankStr}(${
+        unpackedCoordsSnippet}, ${shapeStr}) / 4u];
+      }
+    `;
+  }
+
+  return `
+    fn ${funcName}ByGlobalId(globalId : vec3<u32>) ->f32 {
+      var coords : ${type} = getOutputCoords(globalId);
+      ${coordsSnippet}
+      return f32(${texName}.numbers[getFlatIndex${rankStr}(${
+      unpackedCoordsSnippet}, ${shapeStr})]);
+    }
+
+    fn ${funcName}ByCoords(coordsIn : ${type}) ->f32 {
+      var coords : ${type} = coordsIn;
```

Comment:
Type annotation most likely not needed, should be inferred

Reply:
Yes, removed Type annotation.